interpreter/text/parser.mly

```diff
@@ -39,6 +39,45 @@ let ati i =
 let literal f s =
   try f s with Failure _ -> error s.at "constant out of range"
 
+let range_check i min max at =
+    let i_32 = Int32.to_int i in
+    let min_32 = Int32.of_int min in
+    let max_32 = Int32.of_int max in
+    if i > max_32 || i < min_32 then error at "constant out of range" else i_32
+
+let int8_of_string s = range_check (I32.of_string s.it) (-128) 255 s.at
+let int16_of_string s = range_check (I32.of_string s.it) (-32768) 65535 s.at
+let int32_of_string s = I32.to_bits (literal (fun s -> I32.of_string s.it) s)
+let int64_of_string s = I64.to_bits (literal (fun s -> I64.of_string s.it) s)
+let f32_of_string s = F32.to_bits (literal (fun s -> F32.of_string s.it) s)
+let f64_of_string s = F64.to_bits (literal (fun s -> F64.of_string s.it) s)
+
+let simd_literal f shape ss =
+    let open Bytes in
+    let b = create 16 in
+    (match shape with
+    | "i8x16" ->
+            List.iteri (fun i s -> set_uint8 b i (int8_of_string s)) ss;
+            if List.length ss != 16 then parse_error "unexpected token";
```

Comment:
Instead all the separate conditionals for length tests, you can put guards like "when List.length ss = 2" on the patterns. Then you can unify and move the parse error to the default case. Or more visually, you could even `match shape, ss with` and then have patterns like `| "i64x2", [_; _] ->

Reply:
Good tip, made that change to use guards.